### PR TITLE
Feat/persist node name

### LINKED ISSUE #1522: persist node name across restarts
## Describe the bug

- node currently generates a new random name every time it starts, but it should create a random name when it starts for the first time and use that name every time it starts up afterwards (can write it to the db)

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- On `createDotConfig` check if the node already has a database then try to load the persisted name, if there is no database then try to get the name from **flag** or **toml config file**, if no name was passed by flag or config then generate a random name.
- On node initialization persist the global node name on the database.

## Tests

- Created the test `config.TestGlobalNodeName_WhenNodeAlreadyHasStoredName` where I generate a random name then initialize the node and after calling the `createDotConfig` and check if the returned `dot.Config` has the stored node name or a new one.
- Created the test `dot.TestNode_PersistGlobalName_WhenInitialize` that calls the `dot.InitNode` with a `dot.Config` with a random name to store on the database, and test the created method `LoadGlobalNodeName` that retrieves the stored name from the database.

## Screenshots
- First I initialize the node with the `init` command and pass the `name` flag with the value `mycustomerchainname`, the node was initialized and the name stored.
- Then I call other commands without the name flag and the app can restore the persisted node name from the database.
image
## Issues
closes #1522

### CODE REVIEW

File: chain/gssmr/config.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ key = ""
 unlock = ""
 
 [core]
-roles = 4
+roles = 2
```
Comment (noot) [inline chain/gssmr/config.toml:24]: please change back

File: dot/node.go [hunk 2]
```diff
@@ -104,8 +104,13 @@ func InitNode(cfg *Config) error {
...
 	}
 
+	err = storeGlobalNodeName(cfg.Global.Name, cfg.Global.BasePath)
+	if err != nil {
+		return fmt.Errorf("failed to store global node name: %s", err)
+	}
+
 	logger.Info(
-		"node initialised",
+    "node initialised",
```
Comment (noot) [inline dot/node.go:113]: please run `go fmt ./...`

File: dot/utils.go [hunk 3]
```diff
@@ -235,3 +239,13 @@ func CreateJSONRawFile(bs *BuildSpec, fp string) *os.File {
 	}
 	return WriteConfig(data, fp)
 }
+
+// RandonNodeName generate a new random name
+// if there is no name configured to the node
+func RandonNodeName() string {
```
Comment (noot) [inline dot/utils.go:245]: typo: `RandomNodeName`

File: lib/utils/utils.go [hunk 4]
```diff
@@ -192,3 +195,12 @@ func GetKusamaGenesisPath() string {
 
 	return fp
 }
+
+// RandomNodeName returns a random name when there is no flag --name
+func RandomNodeName() string {
```
Comment (noot) [inline lib/utils/utils.go:200]: looks like this is unused?

**Review by noot — COMMENTED** [review]
it looks like re-initialising the node doesn't update the node name:
```
elizabeth@e:~/go/src/github.com/ChainSafe/gossamer$ ./bin/gossamer --chain gssmr init --force
INFO[04-26|12:07:53] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-26|12:07:53] reinitialising node...                   pkg=cmd basepath=/home/elizabeth/.gossamer/gssmr caller=main.go:321
INFO[04-26|12:07:53] 🕸️ initialising node...             pkg=dot name=remember-pole-46516 id=gssmr basepath=/home/elizabeth/.gossamer/gssmr genesis=./chain/gssmr/genesis.json caller=node.go:59
INFO[04-26|12:07:56] state                                    pkg=state genesis hash=0x98ca3e6f23c6e6d423ed415cde44f690046303e32477eea3f4a8a8273813618b caller=service.go:165
INFO[04-26|12:07:56] node initialised                         pkg=dot name=remember-pole-46516 id=gssmr basepath=/home/elizabeth/.gossamer/gssmr genesis=./chain/gssmr/genesis.json block=0 genesis hash=0x98ca3e6f23c6e6d423ed415cde44f690046303e32477eea3f4a8a8273813618b caller=node.go:112
elizabeth@e:~/go/src/github.com/ChainSafe/gossamer$ ./bin/gossamer --chain gssmr init --force
INFO[04-26|12:07:58] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-26|12:07:58] reinitialising node...                   pkg=cmd basepath=/home/elizabeth/.gossamer/gssmr caller=main.go:321
INFO[04-26|12:07:58] 🕸️ initialising node...             pkg=dot name=remember-pole-46516 id=gssmr basepath=/home/elizabeth/.gossamer/gssmr genesis=./chain/gssmr/genesis.json caller=node.go:59
INFO[04-26|12:08:02] state                                    pkg=state genesis hash=0x98ca3e6f23c6e6d423ed415cde44f690046303e32477eea3f4a8a8273813618b caller=service.go:165
INFO[04-26|12:08:02] node initialised                         pkg=dot name=remember-pole-46516 id=gssmr basepath=/home/elizabeth/.gossamer/gssmr genesis=./chain/gssmr/genesis.json block=0 genesis hash=0x98ca3e6f23c6e6d423ed415cde44f690046303e32477eea3f4a8a8273813618b caller=node.go:112
```

could you update this so that when `gossamer init` is run, a new random name is generated?

it also looks like if I specify a name with `--name`, it isn't used:
```
elizabeth@e:~/go/src/github.com/ChainSafe/gossamer$ ./bin/gossamer --chain gssmr --name my_node
INFO[04-26|12:09:09] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-26|12:09:09] loaded package log configuration         pkg=cmd cfg="{CoreLvl:info SyncLvl:info NetworkLvl:info RPCLvl:info StateLvl:info RuntimeLvl:info BlockProducerLvl:info FinalityGadgetLvl:info}" caller=config.go:130
INFO[04-26|12:09:09] 🕸️ initialising node services...    pkg=dot name=remember-pole-46516 id=gssmr basepath=/home/elizabeth/.gossamer/gssmr caller=node.go:222
```

the `--name` flag should be used instead of what's in the database for that time the node is run. 

could you update these two issues? otherwise, looks good!

**Comment by EclesioMeloJunior** [comment]
Could I follow this priority order?

1. `--name` flag
2. `name` loaded from `config.toml`
3. `name` stored on database

**Comment by noot** [comment]
@EclesioMeloJunior yes, exactly :)

**Comment by EclesioMeloJunior** [comment]
I've fixed the issue with ìnit` command and create the priority order in the node name setup on dot config.

image

File: cmd/gossamer/config.go [hunk 5]
```diff
@@ -444,13 +467,49 @@ func setDotGlobalConfig(ctx *cli.Context, tomlCfg *ctoml.Config, cfg *dot.Global
...
+		return nil
+	}
+
+	// consider the name on config as a second priority
+	if thereIsNameTomlConfig {
+		cfg.Name = tomlCfg.Global.Name
+		return nil
+	}
+
+	// if node was previusly initialised
```
Comment (arijitAD) [inline cmd/gossamer/config.go:491]: typo `previously`
Also, try to remove the new lines. Comments seem to be broken into multiple newlines.

File: cmd/gossamer/config.go [hunk 6]
```diff
@@ -444,13 +467,49 @@ func setDotGlobalConfig(ctx *cli.Context, tomlCfg *ctoml.Config, cfg *dot.Global
...
-		"global configuration",
-		"name", cfg.Name,
-		"id", cfg.ID,
-		"basepath", cfg.BasePath,
-	)
+func setDotGlobalConfigName(ctx *cli.Context, tomlCfg *ctoml.Config, cfg *dot.GlobalConfig) error {
+	globalBasePath := utils.ExpandDir(cfg.BasePath)
+	initialised := dot.NodeInitialized(globalBasePath, false)
+
+	thereIsNameTomlConfig := tomlCfg.Global.Name != ""
```
Comment (arijitAD) [inline cmd/gossamer/config.go:476]: No need to create these variables `thereIsNameTomlConfig` and `thereIsNameFlag`
Directly check the condition in `if-else`

File: cmd/gossamer/config.go [hunk 7]
```diff
@@ -444,13 +467,49 @@ func setDotGlobalConfig(ctx *cli.Context, tomlCfg *ctoml.Config, cfg *dot.Global
...
+	}
+
+	// consider the name on config as a second priority
+	if thereIsNameTomlConfig {
+		cfg.Name = tomlCfg.Global.Name
+		return nil
+	}
+
+	// if node was previusly initialised
+	// and the was not the init command
```
Comment (arijitAD) [inline cmd/gossamer/config.go:492]: Fix comment `the was not the init command`

File: dot/state/db.go [hunk 8]
```diff
@@ -27,6 +27,40 @@ import (
...
+// SetupDatabase will return an instance of database based on basepath
+func SetupDatabase(basepath string) (database.Database, error) {
+	// initialise database using data directory
+	db, err := database.NewBadgerDB(&database.Config{
+		DataDir: basepath,
+	})
+
+	if err != nil {
+		logger.Error(
+			"failed to setup database",
```
Comment (arijitAD) [inline dot/state/db.go:39]: No need for error here. The error will be propagated to the upper level.

**Comment by noot** [comment]
@EclesioMeloJunior everything looks good, except looks like the `--name` flag is no longer used (for both init and the base command), it causes the name to be empty:
```
$ ./bin/gossamer --chain gssmr --name my_node  init --force 
INFO[04-29|10:38:02] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-29|10:38:02] reinitialising node...                   pkg=cmd basepath=/home/elizabeth/.gossamer/gssmr caller=main.go:330
INFO[04-29|10:38:02] 🕸️ initialising node...             pkg=dot name= id=gssmr basepath=/home/elizabeth/.gossamer/gssmr genesis=./chain/gssmr/genesis.json caller=node.go:59
```

```
$ ./bin/gossamer --chain gssmr --key alice --name my_node
INFO[04-29|10:37:18] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-29|10:37:18] loaded package log configuration         pkg=cmd cfg="{CoreLvl:info SyncLvl:info NetworkLvl:info RPCLvl:info StateLvl:info RuntimeLvl:info BlockProducerLvl:info FinalityGadgetLvl:info}" caller=config.go:131
INFO[04-29|10:37:18] 🕸️ initialising node services...    pkg=dot name= id=gssmr basepath=/home/elizabeth/.gossamer/gssmr caller=node.go:224
```

**Comment by EclesioMeloJunior** [comment]
Fixed! When there is the flag we use the flag
```
./bin/gossamer --key alice --chain gssmr --name node_from_brazil                                                        ✹feat/persist-node-name 
INFO[04-29|17:04:56] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-29|17:04:56] loaded package log configuration         pkg=cmd cfg="{CoreLvl:info SyncLvl:info NetworkLvl:info RPCLvl:info StateLvl:info RuntimeLvl:info BlockProducerLvl:info FinalityGadgetLvl:info}" caller=config.go:131
INFO[04-29|17:04:56] 🕸️ initialising node services...    pkg=dot name=node_from_brazil id=gssmr basepath=/home/eclesiomelo/.gossamer/gssmr caller=node.go:224
INFO[04-29|17:04:56] created state service                    pkg=state head=0x64597c55a052d484d9ff357266be326f62573bb4fbdbb3cd49f219396fcebf78 highest number=0 caller=service.go:179
INFO[04-29|17:04:56] Generating p2p identity                  pkg=network RandSeed=0 KeyFile=/home/eclesiomelo

./bin/gossamer account --list --chain gssmr --name othername                                                            ✹feat/persist-node-name 
INFO[04-29|16:21:20] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml
DBUG[04-29|16:21:20] set log configuration                    pkg=cmd --log=info global=info
INFO[04-29|16:21:20] loaded package log configuration         pkg=cmd cfg="{CoreLvl:info SyncLvl:info NetworkLvl:info RPCLvl:info StateLvl:info RuntimeLvl:info BlockProducerLvl:info FinalityGadgetLvl:info}"
DBUG[04-29|16:21:20] global configuration                     pkg=cmd name=othername id=gssmr basepath=~/.gossamer/gssmr
DBUG[04-29|16:21:20] init configuration                       pkg=cmd genesis=./chain/gssmr/genesis.json
```

```
./bin/gossamer init --name beauty_node                                                                           130 ↵  ✹feat/persist-node-name 
Are you sure you want to reinitialise the node? [Y/n]
> y
INFO[04-29|17:11:34] reinitialising node...                   pkg=cmd basepath=/home/eclesiomelo/.gossamer/gssmr caller=main.go:330
INFO[04-29|17:11:34] 🕸️ initialising node...             pkg=dot name=beauty_node id=gssmr basepath=/home/eclesiomelo/.gossamer/gssmr genesis=./chain/gssmr/genesis.json caller=node.go:59
INFO[04-29|17:11:37] state                                    pkg=state genesis hash=0x64597c55a052d484d9ff357266be326f62573bb4fbdbb3cd49f219396fcebf78 caller=initialize.go:134
INFO[04-29|17:11:37] node initialised                         pkg=dot name=beauty_node id=gssmr basepath=/home/eclesiomelo/.gossamer/gssmr genesis=./chain/gssmr/genesis.json block=0 genesis hash=0x64597c55a052d484d9ff357266be326f62573bb4fbdbb3cd49f219396fcebf78 caller=node.go:112

./bin/gossamer --key alice --chain gssmr                                                                                ✹feat/persist-node-name 
INFO[04-29|17:11:43] loading toml configuration...            pkg=cmd config path=./chain/gssmr/config.toml caller=config.go:90
INFO[04-29|17:11:43] loaded package log configuration         pkg=cmd cfg="{CoreLvl:info SyncLvl:info NetworkLvl:info RPCLvl:info StateLvl:info RuntimeLvl:info BlockProducerLvl:info FinalityGadgetLvl:info}" caller=config.go:131
INFO[04-29|17:11:43] 🕸️ initialising node services...    pkg=dot name=beauty_node id=gssmr basepath=/home/eclesiomelo/.gossamer/gssmr caller=node.go:224
INFO[04-29|17:11:43] created state service                    pkg=state head=0x64597c55a052d484d9ff357266be326f62573bb4fbdbb3cd49f219396fcebf78 highest number=0 caller=service.go:179
WARN[04-29|17:11:43] Bootstrap is enabled but no bootstrap nodes are defined pkg=network caller=config.go:134
INFO[04-29|17:11:43] creating runtime...                      pkg=dot interpreter=wasmer caller=services.go:89

```

File: dot/node.go [hunk 9]
```diff
@@ -354,6 +391,34 @@ func setupMetricsServer(address string) {
...
+
+	basestate := state.NewBaseState(db)
+	err = basestate.StoreNodeGlobalName(name)
+	if err != nil {
+		logger.Warn(
+			"failed to store global node name",
+			"basepath", basepath,
+			"error", err,
+		)
+		return nil
```
Comment (timwu20) [inline dot/node.go:409]: why does this return `nil`? Shouldn't the caller of this method be aware if the global node name was stored?

File: dot/node.go [hunk 10]
```diff
@@ -354,6 +391,34 @@ func setupMetricsServer(address string) {
...
+	}
+
+	// close database
+	err = db.Close()
```
Comment (timwu20) [inline dot/node.go:413]: db should be closed whether or not the node global name was stored.  I would use a `defer`.

**Comment by EclesioMeloJunior** [comment]
@timwu20, thank you for point these problems!

File: cmd/gossamer/main.go [hunk 11]
```diff
@@ -28,6 +28,15 @@ import (
 	"github.com/urfave/cli"
 )
 
+const (
```
Comment (noot) [inline cmd/gossamer/main.go:31]: nice :D

**Review by noot — APPROVED** [review]
great work!

**Review by timwu20 — APPROVED** [review]

**Pull request merged**